src/coreclr/vm/codeman.cpp

```diff
@@ -1525,6 +1535,22 @@ void EEJitManager::SetCpuInfo()
 
     if (((cpuFeatures & ARM64IntrinsicConstants_Sve) != 0) && CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnableArm64Sve))
     {
+#if defined(TARGET_LINUX)
+        // prctl() expects vector length in bytes.
+        int maxVectorLength = (maxVectorTBitWidth >> 3);
+
+        // Limit the SVE vector length to 'maxVectorLength' if the underlying hardware offers longer vectors.
+        if ((prctl(PR_SVE_GET_VL, 0,0,0,0) & PR_SVE_VL_LEN_MASK) > maxVectorLength)
```

Comment:
This will set the SVE vector length only for the current thread only. It won't set it for threads that has been started already. Is that correct?

How does it work for new threads? Do new threads inherit SVE vector length of the parent thread or do new threads inherit SVE vector length of the process?

Are we sure that none of the libraries that has been initialized by this point have not cached the vector length?

https://www.man7.org/linux/man-pages/man2/prctl.2.html has explicit warning about use of this API only if you know what you are doing.

It seems that allowing the SVE vector length to be set only before process start is the only reliable option. Setting it here may break all sorts of random things. I am not convinced that we know what we are doing by setting it here.

Reply:
Do you mean in the case of something like the C runtime or in the case of something else hosting the CLR?

Are you thinking the only valid thing for us to do here is to fail to launch for an SVE mismatch (AOT) and to just disable SVE usage otherwise (JIT)?

Reply:
Right, this is obviously not compatible with hosting. (We have no reliable way to tell whether we are hosted.) 

Even without hosting and external libraries outside our control in the picture, there are number of our own threads created in the process by this point (PAL, EventPipe) that will have a wrong size. It is hard to guarantee that none of these threads is going to wonder into managed code.

> Are you thinking the only valid thing for us to do here is to fail to launch for an SVE mismatch (AOT)

I do not see a better option. It suggests that the design we are working with is questionable since it does not work well for AOT.

> just disable SVE usage otherwise (JIT)

Yes, if the JIT is not able to accommodate the SVE length that the process was started with.

Reply:
> If this is only used for debugging/testing and never used in production, then I lean towards it's fine. Anything more then maybe not.

In general these switches are primarily there for debugging/testing purposes. However, they also generally exist as a way to disable or limit intrinsic support if a library is found to have a blocking bug.

It's not great that we can't help setup the process to achieve success, but its also not the end of the world and is something we can ideally give user guidance around.

> Or should DOTNET_MaxVectorTBitWidth be X86 only?

I think it's fine for us to respect it still, that's functionally what AOT compiled for a particular SVE size would have to do after all. It's just a different way to disable SVE support.

Reply:
The documented switches have to be reliable. `DOTNET_MaxVectorTBitWidth` is documented switch. 

It does not sound like that this switch can be reliable. It means that it should have different name, and ideally be a debug-only switch. We do not want to be dealing with inscrutable crashes caused the different parts of the process being configured to different vector sizes.

Reply:
Right. I should have clarified I meant that given your input that we shouldn't take this PR to call `prcrtl` because its unreliable, that the alternative where we simply just don't use SVE if its larger than the `DOTNET_MaxVectorTSize` is fine and still inline with the currently documented behavior for that config switch.

If we provided anything around `prctl` (and it sounds like we're leaning towards `no`), it would need to be a separate undocumented switch, potentially debug only. -- I don't think we have the need to add that given our current testing needs and the known sizes (128 and 256) we'll want to support for existing SVE capable hardware (both consumer and server/cloud).

Reply:
Thanks @tannergooding for clarifying some of the things offline, so it eventually boils down to:

 ```c#
 int VectorTLength = 0;
 int SystemVectorTLength = Get_System_VL();
 
 if (DOTNET_MaxVectorTBitWidth == 0)
 {
    // If we fix getVectorTByteLength() to return system length - then use system provided length
    // Otherwise use 128
    VectorTLength = SystemVectorTLength;
 }
 else if (DOTNET_MaxVectorTBitWidth >= SystemVectorTLength)
 { 
    // For a 256-bit machine, if user provides DOTNET_MaxVectorTBitWidth=512, we will use the
    // maximum available length of 256-bits
    VectorTLength = SystemVectorTLength;
 }
 else if (DOTNET_MaxVectorTBitWidth < SystemVectorTLength)
 {
    // For a 256-bit machine, if user provides DOTNET_MaxVectorTBitWidth=128, we do not want
    // to update it using syscall because that has implications on already initialized components
    // as it might have stale vector length. In that case, disable SVE.
    //
    // If user really wants to downgrade the size, they can call `prctl` or windows `CreateProcess()` 
    // to limit the size before launching dotnet process
    VectorTLength = 128
    DisableSve();
 }
 ```

To implement `Get_System_VL()`, we can use `prctl PR_SVE_GET_VL`, it will be good to use `CNTB` or an equivalent instruction because that way, it will be OS agnostic. We will need that anyway for `getVectorTByteLength()` when `DOTNET_MaxVectorTBitWidth` is not set.

With that said, there is no need to introduce a different environment variable that is `DEBUG` only to support the downgrade scenario.